### ISSUE TITLE
Gradle 9 deprecation warning for ProjectDependency.getDependencyProject()

### ISSUE THREAD

**Issue by nlisker** [beryx/badass-jlink-plugin#287]
Gradle reports that "The ProjectDependency.getDependencyProject() method has been deprecated." in the plugin org.beryx.jlink.
The migration guide points to https://docs.gradle.org/8.11.1/userguide/upgrading_version_8.html#deprecate_get_dependency_project.

This warning shows on Gradle 8.11.1 at the very least.

**Comment by xzel23** [comment]
I fixed this, tested locally, and everything runs without warning. After following the migration guide, a method introduced with Gradle API 8.11 is used, and I am still confused why the compatibility tests for Gradle 7.4 succeed. I now wait for a reply from the Gradle community how this is even possible before merging the change.

**Comment by xzel23** [comment]
Short update: I still didn't find a backwards copatible solution for this. I will look into this again once at least a Gradle 9 RC is published.

**Comment by eskatos** [comment]
You can already try Gradle 9.0 milestones available at https://services.gradle.org/distributions/

**Comment by xzel23** [comment]
Yes, I know. I already tried with a nightly some weeks ago. While I could make it work with the Gradle 9 nightly back then, it wouldn't work with older Gradle versions anymore because they introduced a new way to do things that replaces the old one, but the new API is not available in older versions. I currently don't have the time to come up with a workaround that uses reflection to chase a moving target, so I rather wait for things to settle and at least a beta or RC is available.

From my experience, it takes some time for Gradle to go from RC to GA, and there should be enough time to implement something by then when all the API changes have stabilized.

**Comment by nlisker** [comment]
Yes, there are usually 2 RC versions. You can also ask about this on GitHub or the Gradle Forums, someone usually answers there.

**Comment by xzel23** [comment]
Just asked on Stackoverflow: https://stackoverflow.com/questions/79619019/gradle-replace-deprecated-getdependencyproject-in-a-backwards-compatible-way

**Comment by gwinstanley** [comment]
While I've not tested this on the codebase, I'd have thought something simple like introducing a custom method for `ProjectDependency/Project` resolution would suffice, along these lines:

```diff
diff --git a/src/main/groovy/org/beryx/jlink/util/Util.groovy b/src/main/groovy/org/beryx/jlink/util/Util.groovy
index 5e92fbd7..9146eeda 100644
--- a/src/main/groovy/org/beryx/jlink/util/Util.groovy
+++ b/src/main/groovy/org/beryx/jlink/util/Util.groovy
@@ -342,13 +342,25 @@ class Util {
         if(handledProjects.contains(project)) return []
         handledProjects << project
         def projectDependencies = project.configurations*.dependencies*.withType(ProjectDependency).flatten() as Set<ProjectDependency>
-        List<Project> dependentProjects = projectDependencies*.dependencyProject
+        List<Project> dependentProjects = projectDependencies.collect { getDependencyProject(project, it) }
         dependentProjects.each {
             dependentProjects += getAllDependentProjectsExt(it, handledProjects)
         }
         return dependentProjects.unique()
     }

+    private static boolean GRADLE_VER_GTE_811 = GradleVersion.current().compareTo(GradleVersion.version('8.11'))
+
+    @CompileDynamic
+    private static Project getDependencyProject(Project project, ProjectDependency dep) {
+        if(GRADLE_VER_GTE_811) {
+            return project.project(dep.path)
+        } else {
+            return dep.dependencyProject
+        }
+    }
+
     @CompileDynamic
     static List<String> getDefaultJvmArgs(Project project) {
         try {
```

**Comment by xzel23** [comment]
Looks god. Plugin still does not work on Gradle 9 because of another issue, but at least I am not stuck here anymore.

**Comment by gwinstanley** [comment]
No problem, glad it's helped. It's also possible the `@CompileDynamic` annotation I included could be removed to add a minor performance gain, but I doubt it'd be significant. I left it in more as a safeguard to improve likelihood of working ok, but I can't see any obvious blocker for its removal...worth a try maybe.
